server/controller/src/main/java/ai/starwhale/mlops/domain/panel/PluginService.java

```diff
@@ -0,0 +1,122 @@
+import ai.starwhale.mlops.common.TarFileUtil;
+import ai.starwhale.mlops.common.util.PageUtil;
+import ai.starwhale.mlops.domain.panel.bo.PanelPlugin;
+import ai.starwhale.mlops.domain.panel.mapper.PanelPluginMapper;
+import ai.starwhale.mlops.domain.panel.po.PanelPluginEntity;
+import ai.starwhale.mlops.domain.storage.StoragePathCoordinator;
+import ai.starwhale.mlops.exception.SwProcessException;
+import ai.starwhale.mlops.exception.api.StarwhaleApiException;
+import ai.starwhale.mlops.storage.StorageAccessService;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.pagehelper.PageInfo;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
+
+
+@Slf4j
+@Service
+public class PluginService {
+
+    private final StorageAccessService storageAccessService;
+
+    private final StoragePathCoordinator storagePathCoordinator;
+
+    private final PanelPluginMapper panelPluginMapper;
+
+    private final PanelPluginConvertor panelPluginConvertor;
+
+    private final IdConvertor idConvertor;
+
+    private final ObjectMapper yamlMapper;
+
+    PluginService(
+            StorageAccessService storageAccessService,
+            StoragePathCoordinator storagePathCoordinator,
+            @Qualifier("yamlMapper") ObjectMapper yamlMapper,
+            PanelPluginMapper panelPluginMapper,
+            PanelPluginConvertor panelPluginConvertor,
+            IdConvertor idConvertor
+    ) {
+        this.storageAccessService = storageAccessService;
+        this.storagePathCoordinator = storagePathCoordinator;
+        this.yamlMapper = yamlMapper;
+        this.panelPluginMapper = panelPluginMapper;
+        this.panelPluginConvertor = panelPluginConvertor;
+        this.idConvertor = idConvertor;
+    }
+
+    public void installPlugin(MultipartFile multipartFile) {
+        byte[] content;
+        PanelPlugin.PluginManifest manifest;
+        // 1.save to oss
+        try (var input = multipartFile.getInputStream(); var dupIs = multipartFile.getInputStream()) {
+            content = TarFileUtil.getContentFromTarFile(dupIs, "", "manifest.json");
+            // check manifest
+            if (content == null) {
+                throw new StarwhaleApiException(new SwProcessException(SwProcessException.ErrorType.SYSTEM),
+                    HttpStatus.BAD_REQUEST);
+            }
+            manifest = yamlMapper.readValue(content, PanelPlugin.PluginManifest.class);
+
+            // check if exists
+            var plugin = panelPluginMapper.getByNameAndVersion(manifest.name, manifest.version);
+            if (plugin != null) {
+                throw new StarwhaleApiException(new SwProcessException(SwProcessException.ErrorType.DB),
```

Comment:
this seems to be a validation exception(something wrong with user's input). `SwProcessException` means internal error(something wrong with your code or system) which will cause http code 500.

Reply:
We will response 500 even if I set the http status `HttpStatus.CONFLICT`?

Reply:
> We will response 500 even if I set the http status `HttpStatus.CONFLICT`?

will response `HttpStatus.CONFLICT` here. but SwProcessException is designed to be system error.

Reply:
OK, I get.